File: src/ladym/layers/associative.py
"""L4 — Associative memory (Zettelkasten graph).

Bidirectional edges between memories. Used both for explicit linking (the agent calls
``link()``) and for retrieval-time graph expansion (the deep tier of recall hops along these
edges). Edges have Zep-style temporal validity (``valid_from`` / ``valid_to``) so
contradictions can be retired without being deleted.
"""

from __future__ import annotations

import time
from collections import Counter

from ..schema import Edge
from ..storage.store import SQLiteStore


class AssociativeMemory:
    def __init__(self, store: SQLiteStore):
        self.store = store

    def link(self, src_id: str, dst_id: str, relation: str = "related_to",
             *, weight: float = 1.0, metadata: dict | None = None,
             valid_from: float | None = None, valid_to: float | None = None) -> Edge:
        edge = Edge(
            src_id=src_id,
            relation=relation,
            dst_id=dst_id,
            weight=weight,
            metadata=metadata or {},
            valid_from=valid_from if valid_from is not None else time.time(),
            valid_to=valid_to,
        )
        self.store.put_edge(edge)
        return edge

    def neighbors(self, mem_id: str, *, relation: str | None = None) -> list[Edge]:
        return self.store.neighbors(mem_id, relation=relation)

    def neighbor_counts(self) -> dict[str, int]:
        """Return ``{memory_id: neighbour_count}`` for spreading-activation scoring."""
        rows = self.store.conn.execute(
            "SELECT src_id AS id FROM edges WHERE valid_to IS NULL "
            "UNION ALL "
            "SELECT dst_id AS id FROM edges WHERE valid_to IS NULL"
        ).fetchall()
        c: Counter = Counter(r["id"] for r in rows)
        return dict(c)

    def retire(self, edge_id: str, *, when: float | None = None) -> None:
        """Mark an edge no longer current (sets ``valid_to``)."""
        when = when if when is not None else time.time()
        self.store.conn.execute(
            "UPDATE edges SET valid_to = ? WHERE id = ?", (when, edge_id)
        )
        self.store.conn.commit()

File: layers/associative.go
package layers

import (
	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
)

// AssociativeMemory is L4 — the Zettelkasten graph.
type AssociativeMemory struct {
	Store *storage.SQLiteStore
}

// NewAssociativeMemory builds an AssociativeMemory.
func NewAssociativeMemory(store *storage.SQLiteStore) *AssociativeMemory {
	return &AssociativeMemory{Store: store}
}

// Link creates an edge between two memories. A nil weight defaults to 1.0; an
// explicitly supplied weight (including 0) is stored verbatim.
func (a *AssociativeMemory) Link(srcID, dstID, relation string, weight *float64, metadata map[string]any, validFrom, validTo *float64) (*schema.Edge, error) {
	if relation == "" {
		relation = "related_to"
	}
	w := 1.0
	if weight != nil {
		w = *weight
	}
	if metadata == nil {
		metadata = map[string]any{}
	}
	if validFrom == nil {
		vf := schema.Now()
		validFrom = &vf
	}
	e := &schema.Edge{
		ID:        schema.NewID(),
		SrcID:     srcID,
		Relation:  relation,
		DstID:     dstID,
		Weight:    w,
		ValidFrom: *validFrom,
		ValidTo:   validTo,
		Metadata:  metadata,
	}
	if err := a.Store.PutEdge(e); err != nil {
		return nil, err
	}
	return e, nil
}

// Neighbors returns valid edges touching memID.
func (a *AssociativeMemory) Neighbors(memID, relation string) ([]*schema.Edge, error) {
	return a.Store.Neighbors(memID, relation)
}

// NeighborCounts returns {memory_id: neighbour_count}.
func (a *AssociativeMemory) NeighborCounts() (map[string]int, error) {
	return a.Store.NeighborCounts()
}

// Retire marks an edge no longer current (sets valid_to).
func (a *AssociativeMemory) Retire(edgeID string, when *float64) error {
	t := schema.Now()
	if when != nil {
		t = *when
	}
	_, err := a.Store.DB().Exec("UPDATE edges SET valid_to = ? WHERE id = ?", t, edgeID)
	return err
}
